src/mount/mount-tool.c

```diff
@@ -680,28 +696,194 @@ static int stop_mount(
 }
 
+static int umount_by_mountinfo(sd_bus *bus, const char *what) {
+        _cleanup_fclose_ FILE *proc_self_mountinfo = NULL;
+        int r, r2 = 0;
+
+        proc_self_mountinfo = fopen("/proc/self/mountinfo", "re");
+        if (!proc_self_mountinfo)
+                return -errno;
+
+        for (;;) {
+                _cleanup_free_ char *path = NULL, *where = NULL, *dev = NULL;
+                int k;
+
+                k = fscanf(proc_self_mountinfo,
+                           "%*s "       /* (1) mount id */
+                           "%*s "       /* (2) parent id */
+                           "%*s "       /* (3) major:minor */
+                           "%*s "       /* (4) root */
+                           "%ms "       /* (5) mount point */
+                           "%*s"        /* (6) mount options */
+                           "%*[^-]"     /* (7) optional fields */
+                           "- "         /* (8) separator */
+                           "%*s "       /* (9) file system type */
+                           "%ms"        /* (10) mount source */
+                           "%*s"        /* (11) mount options 2 */
+                           "%*[^\n]",   /* some rubbish at the end */
+                           &path, &dev);
+                if (k != 2) {
+                        if (k == EOF)
+                                break;
+
+                        continue;
+                }
+
+                if (!streq(what, dev))
+                        continue;
+
+                r = cunescape(path, UNESCAPE_RELAX, &where);
+                if (r < 0) {
+                        if (r2 >= 0)
+                                r2 = r;
+                        continue;
+                }
+
+                r = stop_mounts(bus, where, false);
+                if (r < 0) {
+                        if (r2 >= 0)
+                                r2 = r;
+                        continue;
+                }
+        }
+
+        return r2;
+}
+
+static int umount_by_device(sd_bus *bus, const char *what) {
+        _cleanup_udev_device_unref_ struct udev_device *d = NULL;
+        _cleanup_udev_unref_ struct udev *udev = NULL;
+        _cleanup_free_ char *where = NULL;
+        struct stat st;
+        const char *v;
+        int r = 0, r2;
+
+        assert(what);
+
+        if (stat(what, &st) < 0)
+                return log_error_errno(errno, "Can't stat %s: %m", what);
+
+        if (!S_ISBLK(st.st_mode)) {
+                log_error("Path %s is not a block device, don't know what to do.", what);
+                return -ENOTBLK;
+        }
+
+        udev = udev_new();
+        if (!udev)
+                return log_oom();
+
+        d = udev_device_new_from_devnum(udev, 'b', st.st_rdev);
+        if (!d)
+                return log_oom();
+
+        v = udev_device_get_property_value(d, "ID_FS_USAGE");
+        if (!streq_ptr(v, "filesystem")) {
+                log_error("%s does not contain a file system.", what);
```

Comment:
Hm, maybe change that to "does not contain a *known* file system"? We shouldn't categorically state things about which we don't have absolute knowledge.